osp/core/cuds.py

```diff
@@ -200,106 +93,446 @@ def is_a(self, oclass):
         """
         return any(oc in oclass.subclasses for oc in self.oclasses)
 
+    def __getattr__(self, name: str) -> RDFCompatibleType:
+        """Retrieve an attribute whose domain matches the CUDS's oclass.
+
+        Args:
+            name: The name of the attribute.
+
+        Raises:
+            AttributeError: Unknown attribute name.
+
+        Returns:
+            The value of the attribute.
+        """
+        # TODO: The current behavior is to fail with non functional attributes.
+        #  However, the check is based on the amount of values set for an
+        #  attribute and not its definition as functional or non-functional
+        #  in the ontology.
+        # TODO: If an attribute whose domain is not explicitly specified was
+        #  already fixed with __setitem__, then this should also give back
+        #  such attributes (this is backwards compatible).
+        try:
+            if self.session:
+                # Notify the read before getting the attribute by name, as CUDS
+                #  in the deleted buffer may change class to None if done this
+                #  way, making `_get_ontology_attribute_by_name` raise an
+                #  attribute error for such case as it should be.
+                self.session._notify_read(self)
+            attr = self._get_ontology_attribute_by_name(name)
+            values = set(self._attribute_value_generator(attr,
+                                                         _notify_read=False))
+            if len(values) > 1:
+                raise RuntimeError(f"Tried to fetch values of a "
+                                   f"non-functional attribute {attr} using "
+                                   f"the dot notation. This is not "
+                                   f"supported. "
+                                   f"\n \n"
+                                   f"Please use subscript "
+                                   f"notation instead for such attributes: "
+                                   f"my_cuds[{attr}]. This will return a set "
+                                   f"of values instead of a single one")
+            elif len(values) <= 0:
+                return None
+            else:
+                return values.pop()
+        except AttributeError as e:
+            if (  # check if user calls session's methods on wrapper
+                self.is_a(cuba.Wrapper)
+                and self._session is not None
+                and hasattr(self._session, name)
+            ):
+                logger.warning(
+                    "Trying to get non-defined attribute '%s' "
+                    "of wrapper CUDS object '%s'. Will return attribute of "
+                    "its session '%s' instead." % (name, self, self._session)
+                )
+                return getattr(self._session, name)
+            raise AttributeError(name) from e
+
+    def __setattr__(self, name: str,
+                    values: Union[RDFCompatibleType,
```

Comment:
@pablo-de-andres I think if you give `None` to `__setattr__` it deletes the assigned value (the triple is no longer there). Can you check this? In such case, the type hint is wrong and should be

```python
Optional[
                        Union[RDFCompatibleType,
                              Set[RDFCompatibleType]]
```

Reply:
```py
c = city.City(name="Freiburg")
c.name = None
"WARNING 2021-12-10 15:36:29,605 [osp.core]: Type '<class 'NoneType'>' of object None cannot be set as attribute value, as it is incompatible with the OWL standard"
```